docs/proposals/20210112-encryption-roviders.md

```diff
@@ -0,0 +1,112 @@
+
+KubeOne needs to support this feature natively. Meaning the user should be able to enable, disable the feature and rotate keys when needed without having to apply any actions manually. 
+
+## Goals
+
+* Provide a safe path to enable/disable Encryption Providers.
+* Support atomic(?) rotation for existing keys.
+* Rewriting all secret resources (no just secrets) after enable/disable/rotate operations.
+* Support using custom EncryptionProviders configuration file to allow the use of external KMS.
+
+## Non-Goals
+
+* Deploy External KMS.
+* Safely manage (disable/enable/rotate) configuration when a custom configuration file is used. 
+
+## Challenges
+
+The feature has a lot of moving parts; as it requires performing a specific sequence of actions, including changing the KubeAPI configuration, restarting KubeAPI and rewriting all secret resources to apply the encryption. This requires the implementation to be as idempotent as possible with ability to rollback on failure, with out breaking the cluster. 
+
+## Implementation
+
+Unfortunately, it's not possible to simply update the KubeAPI configuration and expect the configuration to reconcile. KubeOne will have to _read_ the _current_ configuration on the custer, _mutate_ it based on the _required_ state and then apply it. Additionally, KubeOne will have to be able to revert changes on any errors and recover safely if the process is interrupted at any point.
+
+The configuration for this will be added under `features` in the KubeOneCluster spec:
+
+```yaml
+apiVersion: kubeone.io/v1beta1
+kind: KubeOneCluster
+features:
+  encryptionProviders:
+    # enable/disable 
+    enabled: true
+    # set to true to automatically rotate keys on apply command. doesn't work with `customProviderFile` set.
+    rotate: true
+    # setting this value will disable automated rotation.
+    customProviderFile: |
+      apiVersion: apiserver.config.k8s.io/v1
+      kind: EncryptionConfiguration
+      resources:
+      - resources:
+        - secrets
+        providers:
+        - identity: {}
+        - aescbc:
+            keys:
+            - name: key1
+            secret: <BASE 64 ENCODED SECRET>
+```
+
+### pre-flight checks
+ * Cluster is healthy.
+ * Current Encryption Providers state/configuration is valid and identical on all control plane nodes.
+ * Ensure Custom Encryption Providers configuration file is valid (if any). 
+
+### Enable Encryption Providers for new cluster
+* Generate a valid configuration file with the `identity` provider set last.
+* Sync the configuration file to all Control Plane nodes. 
+* Set the required KubeAPI configuration and deploy KubeAPI.
+
+### Enable Encryption Providers for existing cluster
+
+* Ensure there is no Encryption Provider Config (manually added by the user, broken previos enable process, etc..) present.
```

Comment:
How will you ensure this?

Reply:
Probe the server to check if the kubeapi has the option enabled.